internal/proxyservice/http/proxy_service.go

```diff
@@ -12,18 +19,269 @@ import (
+	var matchingSubs []Subservice
+	for _, sub := range proxy.subservices {
+		if sub.Matches(r.URL.String()) {
+			matchingSubs = append(matchingSubs, sub)
+		}
+	}
+	return matchingSubs
+}
+
+// selectSubservice finds a subservice matching the request, and issues warnings
+// when there are no matches or more than one match.
+func (proxy *proxyService) selectSubservice(r *gohttp.Request) *Subservice {
+	matchingSubs := proxy.matchingSubservices(r)
+
+	// No match: Warn!
+	if len(matchingSubs) == 0 {
+		msg := "No subservices matched request '%s'"
+		proxy.logger.Warnf(msg, r.URL.String(), matchingSubs)
+		return nil
+	}
+
+	// Multiple matches: Warn!
+	if len(matchingSubs) > 1 {
+		msg := "Multiple subservices matched request '%s': %v\n"
+		proxy.logger.Warnf(msg, r.URL.String(), matchingSubs)
+	}
+
+	// Select first (or only) match
+	subservice := matchingSubs[0]
+	msg := "Using connector '%s' for request %s"
+	proxy.logger.Debugf(msg, subservice.ConnectorID, r.URL.String())
+	return &subservice
+}
+
+
+// ServeHTTP exists to implement the go_http.Handler interface
+func (proxy *proxyService) ServeHTTP(w gohttp.ResponseWriter, r *gohttp.Request) {
+	logger := proxy.logger
+
+	// Log request
+
+	logMsg := "Got request %v %v %v %v"
+	logger.Infoln(logMsg, r.URL.Path, r.Host, r.Method, r.URL.String())
+
+	// Validate request
+
+	if !proxy.validateProxyServerRules(w, r) {
+		return
+	}
+
+	// Remove headers intended only for proxy server.
+
+	r.Header.Del("Proxy-Connection")
+	r.Header.Del("Proxy-Authenticate")
+	r.Header.Del("Proxy-Authorization")
+
+	// Select a subservice
+
+	subservice := proxy.selectSubservice(r)
+
+	// No match: Send request with no authentication.
+	if subservice == nil {
+		proxy.handleRequest(w, r)
+		return
+	}
+
+	// Get current credential values
+
+	creds, err := subservice.RetrieveCredentials()
+	defer internal.ZeroizeCredentials(creds)
+
+	if err != nil {
+		gohttp.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Authenticate request
+
+	err = subservice.Authenticate(r, creds)
+	if err != nil {
+		gohttp.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Send request to target service
+
+	proxy.handleRequest(w, r)
+}
+
+// validateProxyServerRules ensures that the request being made is a valid
+// request for a proxy server.
+func (proxy *proxyService) validateProxyServerRules(
+	w gohttp.ResponseWriter, r *gohttp.Request,
+) bool {
+
+	if r.Method == "CONNECT" {
+		gohttp.Error(w, "CONNECT is not supported.", 405)
+		return false
+	}
+
+	if !r.URL.IsAbs() {
+		errMsg := "This is a proxy server. Non-proxy requests aren't allowed."
+		gohttp.Error(w, errMsg, 500)
+		return false
+	}
+
+	return true
+}
+
+// handleRequest sends the request to the target service and writes the response
+// back to the client.
+func (proxy *proxyService) handleRequest(
+	w gohttp.ResponseWriter, r *gohttp.Request,
+) {
+	logger := proxy.logger
+
+	// TODO: Must be reset for... some reason?
+	r.RequestURI = ""
```

Comment:
I think this is because we're getting our request from the client. I suspect that `RequestURI` is populated from the client...
The Go docs say
> // It is an error to set this field in an HTTP client request.